src/logic/mul.rs

```diff
@@ -0,0 +1,403 @@
+/// 1 bit, it will apply optimizations so that wasted multiplications and additions of zero are
+/// avoided. This function is designed to efficiently handle 5 common kinds of multiplication.
+/// Small here means both small ApInt `BitWidth` and/or small **unsigned** numerical
+/// significance. (Signed multiplication works, but two's complement negative numbers may have a
+/// large number of leading ones, leading to potential inefficiency.)
+/// 
+/// - multiplication of zero by any size integer (no allocation)
+/// - multiplication of small (<= 1 `Digit`) integers (no allocation)
+/// - multiplication of small integers by large integers (or vice versa) (no allocation)
+/// - wrapping multiplication of medium size (<= 512 bits) integers
+/// - multiplication of medium size integers that will not overflow
+/// 
+/// Currently, Karatsuba multiplication is not implemented, so large integer multiplication 
+/// may be very slow compared to other algorithms. According to Wikipedia, Karatsuba algorithms
+/// outperform 𝒪(n^2) algorithms, starting around 320-640 bits
+impl ApInt {
+    /// Multiply-assigns `rhs` to `self` inplace.
+    /// 
+    /// # Errors
+    /// 
+    /// - If `self` and `rhs` have unmatching bitwidths.
+    pub fn wrapping_mul_assign(&mut self, rhs: &ApInt) -> Result<()> {
+        match self.zip_access_data_mut_self(rhs)? {
+            Inl(lhs, rhs) => {
+                *lhs = lhs.wrapping_mul(rhs);
+            }
+            Ext(lhs, rhs) => {
+                //finds the most significant nonzero digit (for later optimizations) and handles
+                //early return of multiplication by zero.
+                let rhs_sig_nonzero: usize = match rhs.iter().rposition(|x| x != &Digit::zero()) {
+                    Some(x) => x,
+                    None => {
+                        for x in lhs.iter_mut() {
+                            x.unset_all()
+                        }
+                        return Ok(());
+                    }
+                };
+                let lhs_sig_nonzero: usize = match lhs.iter().rposition(|x| x != &Digit::zero()) {
+                    Some(x) => x,
+                    None => {
+                        for x in lhs.iter_mut() {
+                            x.unset_all()
+                        }
+                        return Ok(());
+                    }
+                };
+                // For several routines below there was a nested loop that had its first and last
+                // iterations unrolled (and the unrolled loops had their first and last iterations
+                // unrolled), and then some if statements are added for digit overflow checks.
+                // This is done because the compiler probably cannot properly unroll the carry
+                // system, overflow system, and figure out that only `Digit` multiplications were
+                // needed instead of `DoubleDigit` multiplications in some places.
```

Comment:
Have you benchmarked this before going for a manual unroll? Also have you considered to use `crunchy`?

Reply:
This is a different kind of unroll from `crunchy`, but I have made a note for myself.